src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/IMethodSymbolExtensions.cs

```diff
@@ -105,5 +106,15 @@ public static bool IsEntryPoint(this IMethodSymbol methodSymbol, INamedTypeSymbo
                 methodSymbol.ReturnType.SpecialType == SpecialType.System_Int32 ||
                 methodSymbol.ReturnType.OriginalDefinition.Equals(taskType) ||
                 methodSymbol.ReturnType.OriginalDefinition.Equals(genericTaskType));
+
+        /// <summary>
+        /// Tells if an async method returns a task-like type, awaiting for which produces <see langword="void"/> result
+        /// </summary>
+        public static bool IsAsyncEffectivelyReturningVoidTask(this IMethodSymbol method, Compilation compilation)
+        {
+            return method.IsAsync
+                && method.ReturnType is INamedTypeSymbol { Arity: 0 }
+                && method.ReturnType.Equals(compilation.TaskType()) || method.ReturnType.GetAttributes().Any(a => a.AttributeClass?.Equals(compilation.AsyncMethodBuilderAttribute()) ?? false);
```

Comment:
can you break this into individual checks.  In particular, the mix of && and || is very confusing to me.

Thanks!